### PR TITLE
[pkg/ottl] Experiment with Factory Function definition of existing Function

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:** 
This PR is mainly to have a discussion about how OTTL/transformprocessor should support situations where a functionality is needed both as a factory function and as a standard function.  I'm using KeepKeys to center the conversation but there will be other functions/factory functions that fit as well.

(KeepKeys should do regex matching like `DeleteAllMatches` but let's ignore that for now)

At the moment the OTTL has a function called `KeepKeys`, referenced as `keep_keys` in the transform processor, that takes a map and a list of string keys and drops all entries from the map that aren't in the list. 

There is also a need to be able to adjust a map returned from parsing (https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14938 and https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9410).  For this we need a Factory Function so that we can do something like `merge_maps(attributes, KeepKeys(ParseJSON(body), ["log_type"]), "upsert")`.

Both `keep_keys` and `KeepKeys` have the same intention, but `keep_keys` returns `nil` and `KeepKeys` returns the map that was modified.

## Non-exhaustive Options

1. Introduce "keep keys" functionality as a function and a factory function.  The function's implementation can rely on the factory function's.  This is the implementation in this PR.  This solution require's users to understand the difference between a Factory Function and a Function, which is currently only a conceptual difference; nothing in the grammar enforces this (but it could).
2. Update the existing `keep_keys` function to return the value it modifies.  This will allow statements like `merge_maps(attributes, keep_keys(ParseJSON(body), ["log_type"]), "upsert")` but it also allows stuff like `set(attributes["test"], keep_keys(resource.attributes, ["log"]))`.  Since `keep_keys` is not pure, it would allow modifying multiple paths in the same statement, which could get users into bad situations unexpectedly.
3. Remove `keep_keys` and only allow modifying the map with the factory function of `KeepKeys`.  This would require `KeepKeys` to always be used with `set`.  `keep_keys(attributes, ["test"])` would become `set(attributes, KeepKeys(attributes, ["test"]))`.

Personally I like option 1 in combination with a grammar/package change to make factory functions a core concept in the language.  I don't think I want to go to the extreme of removing standard function's ability to return yet. But I do think we can restrict the grammar's value type to functions that start with a capital letter and the Invocation function name to start with a lower case letter.  This starts to enforce a difference between the 2 types of functions (although there is still a problem of the transformprocessor being the one that actually defines the function map, but that can also be adjusted).

### CODE REVIEW

**Review by evan-bradley — COMMENTED** [review]
I like this approach overall, and I agree that some guard rails would help prevent confusion. Is there a way to provide better error messages if we make the two function types distinct within the grammar? I find the grammar's error messages to be pretty cryptic in many cases.

File: pkg/ottl/ottlfuncs/func_factory_keep_keys.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		if val == nil {
+			return nil, nil
+		}
+
+		mapVal, ok := val.(pcommon.Map)
+		if !ok {
+			return nil, fmt.Errorf("target must be a map but got %T", val)
+		}
+
+		mapVal.RemoveIf(func(key string, value pcommon.Value) bool {
```
[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_factory_keep_keys.go:46]: Ya I can try that out.

**Review by MikeGoldsmith — COMMENTED** [review]
I think this looks really good as it both simplifies the logic and enables it be used in other locations. I agree with @evan-bradley's suggestion of the factory itself should not mutate input params 👍🏻

File: pkg/ottl/ottlfuncs/func_factory_keep_keys.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_factory_keep_keys.go]: @evan-bradley @MikeGoldsmith done

**Comment by TylerHelmuth** [comment]
@evan-bradley I'll submit a PR with a grammar change to start enforcing function types

**Comment by evan-bradley** [comment]
>I'll submit a PR with a grammar change to start enforcing function types

If the grammar can't be configured to print better error messages, I'd be alright with enforcing it in the parser instead.

**Comment by TylerHelmuth** [comment]
I am going to close this PR for now in favor of @kentquirk "temp storage" idea.  I'll open a PR for that soon.  We can come back to this PR in the future if needed.

**Pull request closed without merging**